pkg/compactor/block_upload.go

```diff
@@ -0,0 +1,671 @@
+		}
+		l0 := lset[0]
+		for _, l := range lset[1:] {
+			if l.Name < l0.Name {
+				outOfOrderLabels++
+				level.Warn(c.logger).Log("msg",
+					"out-of-order label set: known bug in Prometheus 2.8.0 and below",
+					"labelset", lset.String(),
+					"series", fmt.Sprintf("%d", id),
+				)
+			}
+			l0 = l
+		}
+		if len(chnks) == 0 {
+			return errors.Errorf("empty chunks for series %d", id)
+		}
+
+		if lset.Hash()%shardCount != shard {
+			return errBadRequest{message: fmt.Sprintf("series sharded incorrectly: %s", lset.String())}
+		}
+
+		ooo := 0
+		// Per chunk in series.
+		for i, chnk := range chnks {
+			if i == 0 {
+				continue
+			}
+
+			prev := chnks[i-1]
+
+			// Chunk order within block.
+			if chnk.MinTime > prev.MaxTime {
+				continue
+			}
+
+			if chnk.MinTime == prev.MinTime && chnk.MaxTime == prev.MaxTime {
+				// TODO(bplotka): Calc and check checksum from chunks itself.
+				// The chunks can overlap 1:1 in time, but does not have same data.
+				// We assume same data for simplicity, but it can be a symptom of error.
+				continue
+			}
+
+			// Chunks partly overlaps or out of order.
+			level.Debug(c.logger).Log("msg", "found out of order chunks",
+				"prev_ref", prev.Ref, "next_ref", chnk.Ref,
+				"prev_min_time", timestamp.Time(prev.MinTime).UTC().Format(time.RFC3339Nano),
+				"prev_max_time", timestamp.Time(prev.MaxTime).UTC().Format(time.RFC3339Nano),
+				"next_min_time", timestamp.Time(chnk.MinTime).UTC().Format(time.RFC3339Nano),
+				"next_max_time", timestamp.Time(chnk.MaxTime).UTC().Format(time.RFC3339Nano),
+				"labels", lset, "prev_chunk_index", i-1, "next_chunk_index", i, "chunks_for_series", len(chnks))
+
+			ooo++
+		}
+
+		/*
+			if ooo > 0 {
+				// TODO: Should this be an error?
+					stats.OutOfOrderSeries++
+					stats.OutOfOrderChunks += ooo
+			}
+		*/
+
+		if err := c.verifyChunks(cr, lset, chnks); err != nil {
+			return err
+		}
+	}
+	if ps.Err() != nil {
+		return errors.Wrap(err, "failed iterating over postings")
+	}
+
+	if outOfOrderLabels > 0 {
+		return errBadRequest{message: fmt.Sprintf("block has %d out of order labels", outOfOrderLabels)}
+	}
+
+	return nil
+}
+
+type errBadRequest struct {
+	message string
+}
+
+func (e errBadRequest) Error() string {
+	return e.message
+}
+
+type errBadGateway struct {
+	message string
+	err     error
+}
+
+func (e errBadGateway) Error() string {
+	return e.message
+}
+
+func (e errBadGateway) Unwrap() error {
+	return e.err
+}
+
+func (c *MultitenantCompactor) verifyChunks(cr *chunks.Reader, lset labels.Labels, chnks []chunks.Meta) error {
+	for _, cm := range chnks {
+		ch, err := cr.Chunk(cm.Ref)
+		if err != nil {
+			return errors.Wrapf(err, "failed to read chunk %d", cm.Ref)
+		}
+
+		samples := 0
+		firstSample := true
+		prevTS := int64(-1)
+
+		it := ch.Iterator(nil)
+		for it.Err() == nil && it.Next() {
+			samples++
+			ts, _ := it.At()
+
+			if firstSample {
+				firstSample = false
+				if ts != cm.MinTime {
+					// TODO: Error?
```

Comment:
I'm not sure... technically yes, but I think we can just check if timestamp of the first sample is >= chunk.MinTime, last sample is <= chunk.MaxTime, and timestamps are increasing. We don't need to be strict about exact times, as long as other checks pass.